modules/javafx.graphics/src/main/java/com/sun/prism/j2d/PrismPrintPipeline.java

```diff
@@ -116,17 +119,91 @@ public synchronized ObservableSet<Printer> getAllPrinters() {
-            printerSet =
-                FXCollections.unmodifiableObservableSet
-                   (FXCollections.observableSet(printers));
+            printerSet = FXCollections.observableSet(printers);
+            returnedPrinterSet =
+                FXCollections.unmodifiableObservableSet(printerSet);
+        } else {
+            PrintService[] newServices =
+                    PrintServiceLookup.lookupPrintServices(null, null);
+            if ((newServices.length != printerSet.size()) ||
+                    (lastTime + 120000) > System.currentTimeMillis()) {
+                updatePrinters(newServices);
+            }
+        }
+        lastTime = System.currentTimeMillis();
+        return returnedPrinterSet;
+    }
+
+    private void addNew(PrintService s, Set<Printer> printers) {
+        PrinterImpl impl = new J2DPrinter(s);
+        Printer printer = PrintHelper.createPrinter(impl);
+        impl.setPrinter(printer);
+        printers.add(printer);
+        pMap.put(s, printer);
+    }
+
+    /* Only change a Printer instance if Java 2D changed it.
+     * Otherwise re-map back to the existing Printer instance.
+     * This relies on Java 2D not re-creating a printer too.
+     * Update the existing set so that an app that has cached the printer list
+     * automatically gets the updates. They can also observe changes .. but
+     * if doing so they could see the work in progress, but that's probably
+     * a good thing for an app that is interested.
+     * Two passes -
+     * First pass remove any printers that no longer exist.
+     * Second pass add any new printers.
+     * Finally update the default printer - if needed.
+     */
+    private void updatePrinters(PrintService[] newServices) {
+
+        Set<PrintService> oldServiceSet = pMap.keySet();
+        PrintService[] oldServices = oldServiceSet.toArray(new PrintService[0]);
+
+        for (PrintService os : oldServices) {
+            boolean present = false;
+            for (PrintService ns : newServices) {
+                if (os.equals(ns)) {
+                    present = true;
+                    break;
+                }
+            }
+            if (!present) {
+                Printer printer = pMap.get(os);
+                pMap.remove(os);
+                printerSet.remove(printer);
+            }
+        }
+        for (PrintService s : newServices) {
+            if (!pMap.containsKey(s)) {
+                addNew(s, printerSet);
+            }
+        }
+        PrintService oldDefaultService =
+            (defaultPrinter == null) ? null :
+                ((J2DPrinter)PrintHelper.getPrinterImpl(defaultPrinter)).getService();
+        PrintService newDefaultService = PrintServiceLookup.lookupDefaultPrintService();
+        if (oldDefaultService != null && newDefaultService != null) {
+            if (!oldDefaultService.equals(newDefaultService)) {
+                defaultPrinter = findDefaultPrinter(printerSet, newDefaultService);
```

Comment:
Don't you also want to find the default printer in the case where the old `default` printer was null and the `newDefaultService` is not null?

Reply:
Updating this.